subsys/bluetooth/host/gatt.c

```diff
@@ -2308,6 +2325,17 @@ static int gatt_indicate(struct bt_conn *conn, uint16_t handle,
 		return -EPERM;
 	}
 
+	/* Check if client has subscribed before sending indications, except in
+	 * the case of a service changed indication.
+	 * This is not really required in the Bluetooth specification, but
+	 * follows its spirit.
```

Comment:
Which part of it? Not sending notifications before client is subscribed, or sending service changed notification in any case? 

Should we really always send the service changed notification if the client is not subscribed?

Reply:
> Should we really always send the service changed notification if the client is not subscribed?

It seems so, although it is a bit weird.

> 7.1 SERVICE CHANGED
This Characteristic Value shall be configured to be indicated using the Client
Characteristic Configuration descriptor by a client. Indications caused by
changes to the Service Changed Characteristic Value shall be considered lost
if the client has erroneously not enabled indications in the Client Characteristic
Configuration descriptor (see [Vol 3] Part F, Section 3.3.3).

> 3.3.1 Characteristic declaration
If the server changes any characteristic declaration Attribute Value while the
server has a trusted relationship with any client, then it shall send each client a
Service Changed Indication indicating a change in the service holding the
Characteristic Declaration

Reply:
I read that as we shouldn't send indications when the client hasn't subscribed: 

> Indications caused by changes to the Service Changed Characteristic Value shall be considered lost if the client has erroneously not enabled indications in the Client Characteristic Configuration descriptor

It also seems that we should never attempt to send a _notification_ of the Service Changed characteristic.

Reply:
Yeah it's a bit unclear. At the UPF, when we weren't subscribed microsoft wasn't sending them but google was. So there is definitely some ambiguity in the wording. What is your interpretation @jhedberg ?

Reply:
According to the specification there are no statement that it is not permitted to send a Notification/Indication if the _Client Characteristic Configuration_ bit is not set. It only says that that functionality will not be used.

One of the reasons for that is the peer device might not know what this handle means, he might not even care as he has not enabled the bit's. So it is a waste of energy of the transmitting device.

Reply:
I will remove the exception then.

Reply:
@jori-nordic afaik the original intention was to always force a notification/indication on non-NULL conn, not only because the specification don't mandate proper subscriptions but also because a CCC write could be enqueued/in flight, so my recommendation is to always send the notification/indication if bt_conn is set.

Reply:
so you mean a notification in general or the service changed indication only ? I do get your point for both, but I don't know how a device would react having notifications/indications on attributes they haven't subscribed to.
If that's the behavior we want to keep I can close the PR. @jhedberg @alwa-nordic thoughts ?

Reply:
> I don't know how a device would react having notifications/indications on attributes they haven't subscribed to.

IIRC a device shall discard those. 

In Zephyr we do a lookup on the `conn` and verify the connection handle has been subscribed to:
```c
void bt_gatt_notification(struct bt_conn *conn, uint16_t handle,
			  const void *data, uint16_t length)
{
	struct bt_gatt_subscribe_params *params, *tmp;
	struct gatt_sub *sub;

	BT_DBG("handle 0x%04x length %u", handle, length);

	sub = gatt_sub_find(conn);
	if (!sub) {
		return;
	}

	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&sub->list, params, tmp, node) {
		if (handle != params->value_handle) {
			continue;
		}

		if (check_subscribe_security_level(conn, params)) {
			if (params->notify(conn, params, data, length) ==
			    BT_GATT_ITER_STOP) {
				bt_gatt_unsubscribe(conn, params);
			}
		}
	}
}
```

w.r.t. CCC being in-air/enqueue, that's a bit of stretch IMO. It will be a very rare occurrence that the server have not yet received the CCC write at the same time where the server needs to notify, and even in that case, Zephyr itself would discard the notification. 

I am fine with the check implemented by this PR.

Reply:
I have talked with our spec-man, and I'm leaning slightly towards permitting the direct notifications, but I'm fine with denying it as well.

The wording of the spec does not prohibit notifications/indications to peers that have not subscribed in the CCC. It does not even require the CCC to be present. Spec-wise, the CCC is just a mechanism for soliciting notifications, and in principle could be one of many. E.g., a profile could define a "notification rate" characteristic that would control notifications.

On the other hand, it's quite likely that an application sending notifications unsubscribed peers has a bug. It makes sense to at least log a warning (that can be disabled by configuration) when this happens.

But it's our API, and we can choose to be more restrictive than we have to. So I'm OK either way. Developers that need to implement a profile with a custom subscription mechanism can file a bug report for it when they need it. Maybe it will make sense to expose an unopinionated lower level API then.